### ISSUE TITLE
Training with more than one class

### ISSUE THREAD

**Issue by cthacker** [experiencor/basic-yolo-keras#34]
I think there may be a bug in how the `loss_class` is determined.

I noticed that in preprocessing.py you define the y_batch two different ways:

```
 y_batch = np.zeros((self.config['BATCH_SIZE'], self.config['GRID_H'],
                                    self.config['GRID_W'],  self.config['BOX'], 4+1+1))
 y_batch = np.zeros((self.config['BATCH_SIZE'], self.config['GRID_H'],
                                   self.config['GRID_W'],  self.config['BOX'], 5+self.config['CLASS'])) 
```
which happens to be the same if only have a single class. 

Basically I have been training this model and it doesn't learn classes at all when the data set contains more than one class.

Further edit: I trained the model on a single image (to overfit). I know it can obviously detect this image because the class loss is very low 1e-7. However when I try and predict on it, I get no boxes detected -- so something is wrong here.

EDIT: Ignore the following part, I see that tf.nn.sparse_softmax_cross_entropy_with_logits() works with a single number being given not a one-hot vector

The second part to this is that I'm assuming the label that you assign to y_batch should actually be a vector with 1s corresponding to class index and zero otherwise?

**Comment by experiencor** [comment]
@cthacker Thanks so much for finding out this deadly bug. I changed from tf.nn.softmax_cross_entropy_with_logits() to tf.nn.sparse_softmax_cross_entropy_with_logits(), but missed to change y_batch accordingly. So the second line is wrong. It should be the same as the first line. 

It's strange that it doesn't overfit one image. Can I have one of your image and the corresponding label to reproduce the result?

**Comment by cthacker** [comment]
@experiencor That fixed it! For some reason I convinced myself that the second line was the correct one, so I was using that. 

I'll take a look at the multi-threading, thanks!

**Comment by cthacker** [comment]
Well this was premature I guess. 

I'm still having issues training (and even overfitting correctly). When I'm training loss_conf really takes forever to come down, so I think something is going on there.

If you want to try and reproduce, I'm using TinyYolo 416x416.

Additionally I noticed that in preprocessing we use center_x, y, w, h and they are all scaled to be on the grid, but draw_boxes actually doesn't take into account the grid so you should be dividing by the grid size for each xmin, xmax, ymin, ymax etc if you want to draw one of those boxes.

I'm thinking that might be messing with the bbox_iou score as well? But yeah even when I overfit my pred_box_conf is really not that high ... something is going on there.

**Comment by cthacker** [comment]
Following up on this more: when overfitting with a single image I know 
`y_true[0, 7, 6, 3, 4] = 1` and at the end of one epoch `y_pred[0, 7, 6, 3, 4] = 1.76` (or 0.85 after sigmoid)
However when I run predict `netout[7, 6, 3, 4] = -4.08` (or ~.017 after signmoid)

Am I correct in thinking these should be the same for the same input image? Is it possible for you to test training a model and seeing if you get similar results?

**Comment by experiencor** [comment]
@cthacker These may not be the same value if you use augmentations when training and validating. Can you try to turn off the jitter parameter of ``train_batch = BatchGenerator(train_imgs, ...)``` in line 397 of frontend.py?

**Comment by cthacker** [comment]
I modified your code quite a bit, but I have turned off aumentation. I'm going to try a fresh checkout and see what happens.

**Comment by cthacker** [comment]
@experiencor, 

With a fresh checkout starting with tiny_yolo_features.h5, I cannot overfit a single image on the racoon data set. Are you able to?

**Comment by cthacker** [comment]
@experiencor I have a fresh checkout and am using tiny_yolo_features.h5 as my starting point. I am not able to overfit a single image of the racoon dataset. I have jitter and shuffling turned off and am fitting on racoon-1.jpg (validation data is the exact same). 

Are you able to do this?

**Comment by cthacker** [comment]
@experiencor 

Have you been able to train on multliple classes?
